### PR TITLE
[DependencyInjection] [Routing] [Config] Recursive directory loading

### PR DESCRIPTION
**Pull request by lavoiesl**
As asked by @stof, this PR is a remake of #11059 on the 2.7 branch.

| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #11045, #11059 |
| License | MIT |
| Doc PR | TODO |

See those about creating a cookbook for better file organization:
Issue: symfony/symfony-docs#3680
PR: symfony/symfony-docs#3885

This PR allows to specify a directory to load and it will recursively load all files from it, using the `LoaderResolver`.
### Example

``` yml
# app/config/config.yml
imports:
    - { resource: 'dev/' }
```

```
└── app
   ├── config
   │  ├── common
   │  │  └── framework.yml
   │  │  └── assetic.yml
   │  └── config.yml
```

It also works for routing configuration:

``` yml
_common:
    resource: "common/"
    type: directory
```
### Advantages
1. Allows easy separation of configuration files per bundle.
2. This could eventually allow a bundle installation tool that would automatically install needed configuration files.
3. Could superseed the `AnnotationDirectoryLoader`, so it does not duplicate the logic.
4. Is entirely optional and introduces no BC break.
5. Files are loaded lexically with scandir, allowing someone to control the load order through a number prefix.
6. Even if the initial configuration load may be a little bit longer, it does not impact the compiled configuration (production).
### Graceful degradation

~~In [`FrameworkBundle`](https://github.com/lavoiesl/symfony/blob/f7f5dd7781b9ed9b84f65e8e28161c8454f01c29/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/RoutingResolverPass.php#L34) and [`HttpKernel`](https://github.com/lavoiesl/symfony/blob/f7f5dd7781b9ed9b84f65e8e28161c8454f01c29/src/Symfony/Component/HttpKernel/Kernel.php#L748), the `DirectoryLoader` is only added if available. This means you can have `FrameworkBundle` `2.7` with `Routing` and `DependencyInjection` `2.6` and everything will work, except the directory loading.~~ As asked by @stof, the version requirements are now 2.7 for both.
### Conflicts

For now, the Routing DirectoryLoader requires the type `directory` to be specified so it does not conflict with `AnnotationDirectoryLoader`. However, this could be refactored.

~~The type declaration requires a [patch in AsseticBundle](https://github.com/symfony/AsseticBundle/commit/f76291c4655f6b7b29f4d20650c020d3be516ad5) for it to be used on `framework.router.resource`. The patch is already merged in master, but not in a release yet.~~ Released in [2.5.0](https://github.com/symfony/AsseticBundle/releases/tag/v2.5.0). Changing the main router would be like this: 

``` yml
framework:
    router:
        resource: "%kernel.root_dir%/config/routing/common/"
        type: directory
```

### CODE REVIEW

**Comment by lavoiesl** [comment]
As of now, some Travis tests fail for a reason I can’t identify. The tests run fine on my computer.

**Comment by nicolas-grekas** [comment]
To reproduce on your local computer:
`cd src/Symfony/Component/...`
`composer up` for "high" tests or `composer up --prefer-lowest --prefer-stable` "low" ones
`phpunit`

**Comment by lavoiesl** [comment]
@nicolas-grekas And do you know what could be causing the issue? AFAIK, the class is in the proper place. Isn’t the autoloader supposed to do its job?

**Comment by lavoiesl** [comment]
Oh, nevermind, it is actually pretty simple: it downloads the `symfony/dependencyinjection` which hasn’t the class yet. So this is expected to fail and will be resolved when merged.

**Comment by nicolas-grekas** [comment]
You're right for "high" tests. But merging won't solve "low" tests unless you state that ~2.7 is the minimum allowed version (in components that relies on your newly added class). One or more composer.json need an update

File: src/Symfony/Component/DependencyInjection/Loader/DirectoryLoader.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+     * Returns true if this class supports the given resource.
+     *
+     * @param mixed  $resource A resource
+     * @param string $type     The resource type
+     *
+     * @return bool true if this class supports the given resource, false otherwise
+     */
+    public function supports($resource, $type = null)
+    {
+        return 'directory' === $type || (!$type && preg_match('/\/$/', $resource) === 1); // ends with a slash
```
Comment (aitboudad) [inline src/Symfony/Component/DependencyInjection/Loader/DirectoryLoader.php]: I'd prefer to check is_dir instead of is ended with a slash

File: src/Symfony/Component/Routing/Loader/DirectoryLoader.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+use Symfony\Component\Config\Loader\FileLoader;
+use Symfony\Component\Routing\RouteCollection;
+use Symfony\Component\Config\Resource\DirectoryResource;
+
+class DirectoryLoader extends FileLoader
+{
+    private $currentDir;
+
+    /**
+     * @param mixed  $file The resource
```
Comment (aitboudad) [inline src/Symfony/Component/Routing/Loader/DirectoryLoader.php]: just use {@inheritdoc}

File: src/Symfony/Component/Routing/Loader/DirectoryLoader.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
+<?php
+
```
Comment (aitboudad) [inline src/Symfony/Component/Routing/Loader/DirectoryLoader.php]: licence header is missing

File: src/Symfony/Component/Routing/Loader/DirectoryLoader.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    /**
+     * Returns true if this class supports the given resource.
+     *
+     * @param mixed  $resource A resource
+     * @param string $type     The resource type
+     *
+     * @return bool true if this class supports the given resource, false otherwise
+     */
+    public function supports($resource, $type = null)
```
Comment (aitboudad) [inline src/Symfony/Component/Routing/Loader/DirectoryLoader.php]: use {@inheritdoc}

File: src/Symfony/Component/Routing/Loader/DirectoryLoader.php (hunk 3 above)
Comment (lavoiesl) [inline src/Symfony/Component/Routing/Loader/DirectoryLoader.php]: fixed

File: src/Symfony/Component/Routing/Loader/DirectoryLoader.php (hunk 2 above)
Comment (lavoiesl) [inline src/Symfony/Component/Routing/Loader/DirectoryLoader.php]: Fixed

File: src/Symfony/Component/DependencyInjection/Loader/DirectoryLoader.php (hunk 1 above)
Comment (lavoiesl) [inline src/Symfony/Component/DependencyInjection/Loader/DirectoryLoader.php]: After extensive testing, I had to remove `is_dir` checks in both loaders because `->supports()` is not aware of the current location. Therefore, trailing slash is mandatory. I also updated the tests to reflect this.

File: src/Symfony/Component/Routing/Loader/DirectoryLoader.php (hunk 4 above)
Comment (lavoiesl) [inline src/Symfony/Component/Routing/Loader/DirectoryLoader.php]: Fixed

**Comment by aitboudad** [comment]
:+1:

**Comment by lavoiesl** [comment]
The errors remaining in Travis are not related to this PR.

@nicolas-grekas I added graceful degradation instead of bumping the requirements. See the updated description.

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 5]
```diff
@@ -736,13 +737,20 @@ protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container
...
             new ClosureLoader($container),
-        ));
+        );
+
+        // DirectoryLoader only appears in DI 2.7
+        if (class_exists('Symfony\\Component\\DependencyInjection\\Loader\\DirectoryLoader')) {
+            $loaders[] = new DirectoryLoader($container, $locator);
+        }
+
+        $resolver = new LoaderResolver($loaders);
```
Comment (lavoiesl) [inline src/Symfony/Component/HttpKernel/Kernel.php]: I am wondering why the default sets of loaders is written in `HttpKernel`, shouldn’t there be something like `Symfony\Component\DependencyInjection\Loader\LoaderResolver` that would contain the set of default loaders? This is out of scope of this PR though.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/RoutingResolverPass.php [hunk 6]
```diff
@@ -30,6 +30,12 @@ public function process(ContainerBuilder $container)
 
         $definition = $container->getDefinition('routing.resolver');
 
+        // routing.loader.directory only appears in Routing 2.7
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/RoutingResolverPass.php]: I would rather bump the requirement, so that people using FrameworkBundle 2.7 can be sure about which loaders are supported

File: src/Symfony/Component/DependencyInjection/Loader/DirectoryLoader.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+     */
+    public function load($file, $type = null)
+    {
+        $file = rtrim($file, '/');
+        $path = $this->locator->locate($file);
+        $this->container->addResource(new DirectoryResource($path));
+
+        foreach (scandir($path) as $dir) {
+            if ($dir[0] !== '.') {
+                if (is_dir("$path/$dir")) {
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Loader/DirectoryLoader.php]: please use concatenation rather than interpolation

File: src/Symfony/Component/DependencyInjection/Loader/DirectoryLoader.php [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports($resource, $type = null)
+    {
+        if ('directory' === $type) {
+            return true;
+        } elseif (null === $type) {
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Loader/DirectoryLoader.php]: just use ``if`` as the previous ``if`` returns

File: src/Symfony/Component/DependencyInjection/Tests/Loader/DirectoryLoaderTest.php [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+    public function testImports()
+    {
+        $this->loader->resolve('directory/import/import.yml')->load('directory/import/import.yml');
+        $this->assertEquals(array('ini' => 'ini', 'yaml' => 'yaml'), $this->container->getParameterBag()->all(), '->load() takes a single file that imports a directory');
+    }
+
+    /**
+     * @covers Symfony\Component\DependencyInjection\Loader\DirectoryLoader::__construct
+     * @covers Symfony\Component\DependencyInjection\Loader\DirectoryLoader::load
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Tests/Loader/DirectoryLoaderTest.php]: we don't use ``@covers`` in Symfony

File: src/Symfony/Component/DependencyInjection/Tests/Loader/DirectoryLoaderTest.php [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+use Symfony\Component\DependencyInjection\Loader\DirectoryLoader;
+use Symfony\Component\Config\Loader\LoaderResolver;
+use Symfony\Component\Config\FileLocator;
+
+class DirectoryLoaderTest extends \PHPUnit_Framework_TestCase
+{
+    protected static $fixturesPath;
+
+    protected $container;
+    protected $loader;
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Tests/Loader/DirectoryLoaderTest.php]: please make them private

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 11]
```diff
@@ -736,13 +737,20 @@ protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container
...
             new XmlFileLoader($container, $locator),
             new YamlFileLoader($container, $locator),
             new IniFileLoader($container, $locator),
             new PhpFileLoader($container, $locator),
             new ClosureLoader($container),
-        ));
+        );
+
+        // DirectoryLoader only appears in DI 2.7
+        if (class_exists('Symfony\\Component\\DependencyInjection\\Loader\\DirectoryLoader')) {
```
Comment (stof) [inline src/Symfony/Component/HttpKernel/Kernel.php]: I would prefer bumping the dependency and defining it all the time. Much better DX.

File: src/Symfony/Component/Routing/Loader/DirectoryLoader.php [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+        $collection = new RouteCollection();
+        $collection->addResource(new DirectoryResource($path));
+
+        foreach (scandir($path) as $dir) {
+            if ($dir[0] !== '.') {
+                $this->setCurrentDir($path);
+
+                $subType = is_dir("$path/$dir") ? 'directory' : null;
+                $subCollection = $this->import("$path/$dir", $subType, false, $path);
```
Comment (stof) [inline src/Symfony/Component/Routing/Loader/DirectoryLoader.php]: use concatenation rather than interpolation

File: src/Symfony/Component/Routing/Loader/DirectoryLoader.php [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    public function load($file, $type = null)
+    {
+        $path = $this->locator->locate($file);
+
+        $collection = new RouteCollection();
+        $collection->addResource(new DirectoryResource($path));
+
+        foreach (scandir($path) as $dir) {
+            if ($dir[0] !== '.') {
+                $this->setCurrentDir($path);
```
Comment (stof) [inline src/Symfony/Component/Routing/Loader/DirectoryLoader.php]: don't you need to add the trailing / to allow recursion, as done in the DI component ?

File: src/Symfony/Component/Routing/Tests/Fixtures/directory_import/import.yml [hunk 14]
```diff
@@ -0,0 +1,3 @@
+_directory:
+    resource: "../directory"
+    type: directory
```
Comment (stof) [inline src/Symfony/Component/Routing/Tests/Fixtures/directory_import/import.yml]: please add the final EOL

File: src/Symfony/Component/Routing/Tests/Loader/DirectoryLoaderTest.php [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+use Symfony\Component\Routing\Loader\YamlFileLoader;
+use Symfony\Component\Routing\Loader\AnnotationFileLoader;
+use Symfony\Component\Config\Loader\LoaderResolver;
+use Symfony\Component\Config\FileLocator;
+use Symfony\Component\Routing\RouteCollection;
+
+class DirectoryLoaderTest extends AbstractAnnotationLoaderTest
+{
+    protected $loader;
+    protected $reader;
```
Comment (stof) [inline src/Symfony/Component/Routing/Tests/Loader/DirectoryLoaderTest.php]: should be private

File: src/Symfony/Component/Routing/Tests/Loader/DirectoryLoaderTest.php [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+    protected function setUp()
+    {
+        parent::setUp();
+
+        $locator      = new FileLocator();
+        $this->reader = $this->getReader();
+        $this->loader = new DirectoryLoader($locator);
+        $resolver     = new LoaderResolver(array(
```
Comment (stof) [inline src/Symfony/Component/Routing/Tests/Loader/DirectoryLoaderTest.php]: please don't align ``=`` operators

File: src/Symfony/Component/Routing/Tests/Loader/DirectoryLoaderTest.php [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+        $this->verifyCollection($collection);
+    }
+
+    public function testImportDirectory()
+    {
+        $collection = $this->loader->load(__DIR__.'/../Fixtures/directory_import', 'directory');
+        $this->verifyCollection($collection);
+    }
+
+    protected function verifyCollection(RouteCollection $collection)
```
Comment (stof) [inline src/Symfony/Component/Routing/Tests/Loader/DirectoryLoaderTest.php]: should be private

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 11 above)
Comment (lavoiesl) [inline src/Symfony/Component/HttpKernel/Kernel.php]: Are you sure about that? Most of the time, you will have latest version, but if you do require an old version, it’ll still work, just not that part. Seems like a rather small feature to have such a big version bump.

File: src/Symfony/Component/Routing/Loader/DirectoryLoader.php (hunk 13 above)
Comment (lavoiesl) [inline src/Symfony/Component/Routing/Loader/DirectoryLoader.php]: Good catch, I will change it for consistency, but actually no, the Routing DirectoryLoader only activates (and always activates) when the type is set to `directory`.

**Comment by lavoiesl** [comment]
Errors in Travis not related to my PR.

File: src/Symfony/Component/HttpKernel/Kernel.php (hunk 11 above)
Comment (jakzal) [inline src/Symfony/Component/HttpKernel/Kernel.php]: @laurentb I don't think it's worth it. People using Standard Edition will get both components in right versions. Anyone using DependencyInjection and HttpKernel components standalone, is already maintaining versions himself in his composer.json. I fail to see a scenario when he'd like to bump one component's version bot not the other.

**Comment by webda2l** [comment]
Too late for 2.7? Will be a nice addition.

**Comment by nicolas-grekas** [comment]
Cherry-picked in #14700 thank you @lavoiesl!
